marketplace-service/src/main/java/com/axonivy/market/service/impl/FileDownloadServiceImpl.java

```diff
@@ -0,0 +1,182 @@
+package com.axonivy.market.service.impl;
+
+import com.axonivy.market.entity.Metadata;
+import com.axonivy.market.service.FileDownloadService;
+import org.springframework.stereotype.Service;
+
+import lombok.extern.log4j.Log4j2;
+import org.apache.commons.lang3.SystemUtils;
+import org.springframework.web.client.RestTemplate;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.FileAttribute;
+import java.nio.file.attribute.PosixFilePermission;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.util.Comparator;
+import java.util.EnumSet;
+import java.util.Enumeration;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Stream;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+@Service
+@Log4j2
+public class FileDownloadServiceImpl implements FileDownloadService {
+  private static final String ZIP_EXTENSION = ".zip";
+  private static final Set<PosixFilePermission> PERMS = EnumSet.allOf(PosixFilePermission.class);
+  private static final int THRESHOLD_ENTRIES = 10000;
+  private static final int THRESHOLD_SIZE = 1000000000;
+  private static final double THRESHOLD_RATIO = 10;
+
+  private byte[] downloadFileByRestTemplate(String url) {
+    return new RestTemplate().getForObject(url, byte[].class);
+  }
+
+  @Override
+  public String downloadAndUnzipProductContentFile(String url, Metadata snapShotMetadata) throws IOException {
+    String unzippedFilePath = String.join(File.separator, ROOT_STORAGE_FOR_PRODUCT_CONTENT,
+        snapShotMetadata.getArtifactId());
+    createFolder(unzippedFilePath);
+
+    Path tempZipPath = createTempFile();
+    Files.write(tempZipPath, downloadFileByRestTemplate(url));
+
+    unzipFile(tempZipPath.toString(), unzippedFilePath);
+
+    grantNecessaryPermissionsFor(unzippedFilePath);
```

Comment:
we does not sharing this folder to other use like nginx or www-data, only tomcat user create and he delete it, so we don't need to grant the permission, just keep the default